File: web_tutor/static/realtime-guide.js
/**
 * 實時引導系統
 * 在用戶輸入過程中提供智能引導和提示
 */

class RealtimeGuide {
    constructor(lesson = null) {
        this.lesson = lesson;
        this.codeRequirements = {};
        this.typingTimer = null;
        this.lastCode = '';
        this.guideLevel = 'moderate'; // 'minimal', 'moderate', 'helpful'
        this.hintDelay = 2000; // 2秒後顯示提示
        this.suggestionContainer = null;
        
        if (lesson && lesson.validator) {
            this.codeRequirements = lesson.validator.code_requirements || {};
        }
    }
    
    /**
     * 初始化引導系統
     */
    init(editorElement, suggestionContainer) {
        this.editor = editorElement;
        this.suggestionContainer = suggestionContainer;
        this.setupEventListeners();
    }
    
    /**
     * 設置事件監聽器
     */
    setupEventListeners() {
        if (!this.editor) return;
        
        let typingTimeout;
        let lastTypingTime = Date.now();
        
        this.editor.addEventListener('input', (e) => {
            const now = Date.now();
            const code = this.editor.value;
            const cursorPos = this.getCursorPosition();
            
            // 清除之前的計時器
            clearTimeout(typingTimeout);
            
            // 如果停止輸入一段時間，提供引導
            typingTimeout = setTimeout(() => {
                this.provideGuidance(code, cursorPos);
            }, this.hintDelay);
            
            // 實時語法檢查（輕量級）
            this.checkSyntaxRealtime(code, cursorPos);
            
            this.lastCode = code;
            lastTypingTime = now;
        });
        
            // 當游標移動時，更新提示（只在有課程要求時）
        this.editor.addEventListener('keyup', (e) => {
            if (e.key === 'ArrowUp' || e.key === 'ArrowDown' || 
                e.key === 'ArrowLeft' || e.key === 'ArrowRight') {
                // 只在有課程要求時才更新上下文提示
                if (this.codeRequirements && Object.keys(this.codeRequirements).length > 0) {
                    const code = this.editor.value;
                    const cursorPos = this.getCursorPosition();
                    this.updateContextualHint(code, cursorPos);
                }
            }
        });
    }
    
    /**
     * 獲取游標位置
     */
    getCursorPosition() {
        if (!this.editor) return { line: 1, col: 1 };
        
        const text = this.editor.value;
        const cursorPos = this.editor.selectionStart;
        const textBeforeCursor = text.substring(0, cursorPos);
        const lines = textBeforeCursor.split('\n');
        
        return {
            line: lines.length,
            col: lines[lines.length - 1].length + 1
        };
    }
    
    /**
     * 實時語法檢查（輕量級）
     */
    checkSyntaxRealtime(code, cursorPos) {
        if (!code.trim()) return;
        
        // 簡單的語法檢查
        const suggestions = [];
        
        // 檢查常見錯誤
        const lines = code.split('\n');
        const currentLine = lines[cursorPos.line - 1] || '';
        
        // 檢查縮排問題
        if (currentLine.trim()) {
            const prevLine = lines[cursorPos.line - 2];
            if (prevLine && prevLine.trim().endsWith(':') && 
                !currentLine.match(/^\s+/)) {
                suggestions.push({
                    type: 'indentation',
                    message: '💡 上一行以冒號結尾，這一行需要縮排',
                    severity: 'info'
                });
            }
        }
        
        // 檢查常見拼寫錯誤
        const typos = this.checkTypos(currentLine);
        suggestions.push(...typos);
        
        this.showSuggestions(suggestions, 'realtime');
    }
    
    /**
     * 檢查拼寫錯誤
     */
    checkTypos(line) {
        const suggestions = [];
        const commonMistakes = {
            'prnt': 'print',
            'prin': 'print',
            'rng': 'range',
            'rge': 'range',
            'fr ': 'for ',
            'whle': 'while',
            'whil': 'while'
        };
        
        for (const [mistake, correct] of Object.entries(commonMistakes)) {
            if (line.toLowerCase().includes(mistake) && 
                !line.toLowerCase().includes(correct)) {
                suggestions.push({
                    type: 'typo',
                    message: `💡 是否想輸入 '${correct}'？`,
                    severity: 'hint'
                });
                break;
            }
        }
        
        return suggestions;
    }
    
    /**
     * 提供引導
     */
    provideGuidance(code, cursorPos) {
        if (!code.trim() || code.length < 10) return;
        
        // 如果課程沒有 code_requirements，就不提供結構性引導
        if (!this.codeRequirements || Object.keys(this.codeRequirements).length === 0) {
            return;
        }
        
        const suggestions = [];
        const hints = [];
        const warnings = [];
        
        // 根據課程要求提供引導
        if (this.codeRequirements.requires_loop) {
            const hasLoop = this.hasLoop(code);
            if (!hasLoop) {
                const loopType = this.codeRequirements.loop_type;
                if (loopType === 'for') {
                    hints.push({
                        type: 'missing_loop',
                        message: '💡 提示：此題目要求使用 for 循環',
                        suggestion: '嘗試使用：for i in range(...):',
                        severity: 'info'
                    });
                } else if (loopType === 'while') {
                    hints.push({
                        type: 'missing_loop',
                        message: '💡 提示：此題目要求使用 while 循環',
                        suggestion: '嘗試使用：while 條件:',
                        severity: 'info'
                    });
                } else {
                    hints.push({
                        type: 'missing_loop',
                        message: '💡 提示：此題目要求使用循環',
                        suggestion: '可以使用 for 或 while 循環',
                        severity: 'info'
                    });
                }
            }
        }
        
        // 檢查硬編碼
        if (this.codeRequirements.forbids_hardcode) {
            if (this.detectHardcode(code)) {
                warnings.push({
                    type: 'hardcode',
                    message: '⚠️ 檢測到可能的硬編碼寫法',
                    suggestion: '考慮使用循環來簡化代碼',
                    severity: 'warning'
                });
            }
        }
        
        // 檢查函數
        if (this.codeRequirements.requires_function) {
            if (!code.includes('def ')) {
                const funcName = this.codeRequirements.function_name;
                hints.push({
                    type: 'missing_function',
                    message: funcName 
                        ? `💡 提示：需要定義名為 '${funcName}' 的函數`
                        : '💡 提示：此題目要求定義函數',
                    suggestion: funcName 
                        ? `def ${funcName}():`
                        : '使用 def 函數名稱(): 來定義函數',
                    severity: 'info'
                });
            }
        }
        
        // 檢查 if 語句
        if (this.codeRequirements.requires_if) {
            if (!code.includes('if ')) {
                hints.push({
                    type: 'missing_if',
                    message: '💡 提示：此題目要求使用 if 語句',
                    suggestion: '使用 if 條件: 來進行條件判斷',
                    severity: 'info'
                });
            }
        }
        
        // 根據幫助級別過濾提示
        const filteredHints = this.filterByGuideLevel(hints);
        
        this.showSuggestions([...warnings, ...filteredHints], 'guidance');
    }
    
    /**
     * 更新上下文提示
     */
    updateContextualHint(code, cursorPos) {
        // 只在有課程要求時提供上下文提示，避免在不該提示的時候提示
        if (!this.codeRequirements || Object.keys(this.codeRequirements).length === 0) {
            return;
        }
        
        const lines = code.split('\n');
        const currentLine = lines[cursorPos.line - 1] || '';
        
        let hint = null;
        
        // 如果課程要求使用 for 循環，且正在輸入 for
        if (this.codeRequirements.requires_loop && 
            (this.codeRequirements.loop_type === 'for' || !this.codeRequirements.loop_type)) {
            if (currentLine.includes('for') && 
                !currentLine.includes('in') && 
                !currentLine.includes('range')) {
                hint = {
                    type: 'contextual',
                    message: '💡 提示：for 循環通常搭配 range() 使用',
                    example: 'for i in range(5):',
                    severity: 'hint'
                };
            }
        }
        
        // 如果課程要求定義函數，且正在輸入函數定義
        if (this.codeRequirements.requires_function) {
            if (currentLine.includes('def') && !currentLine.includes(':')) {
                hint = {
                    type: 'contextual',
                    message: '💡 提示：函數定義需要以冒號 : 結尾',
                    severity: 'hint'
                };
            }
        }
        
        // 如果課程要求使用 if 語句，且正在輸入 if
        if (this.codeRequirements.requires_if) {
            if (currentLine.includes('if') && 
                 !currentLine.includes(':') && 
                 !currentLine.trim().endsWith(':')) {
                hint = {
                    type: 'contextual',
                    message: '💡 提示：if 語句需要以冒號 : 結尾',
                    severity: 'hint'
                };
            }
        }
        
        if (hint) {
            this.showSuggestions([hint], 'contextual');
        }
    }
    
    /**
     * 檢查是否有循環
     */
    hasLoop(code) {
        return /for\s+\w+\s+in/.test(code) || /while\s+/.test(code);
    }
    
    /**
     * 檢測硬編碼
     * 只在課程明確禁止硬編碼時才檢測
     */
    detectHardcode(code) {
        // 如果課程沒有禁止硬編碼的要求，就不檢測
        if (!this.codeRequirements || !this.codeRequirements.forbids_hardcode) {
            return false;
        }
        
        const printMatches = code.match(/print\(/g);
        const printCount = printMatches ? printMatches.length : 0;
        return printCount > 3 && !this.hasLoop(code);
    }
    
    /**
     * 根據幫助級別過濾提示
     */
    filterByGuideLevel(hints) {
        if (this.guideLevel === 'minimal') {
            // 只顯示關鍵提示
            return hints.filter(h => h.severity === 'warning' || h.type === 'missing_loop');
        } else if (this.guideLevel === 'moderate') {
            // 顯示重要提示
            return hints.filter(h => h.severity !== 'hint');
        } else {
            // 顯示所有提示
            return hints;
        }
    }
    
    /**
     * 顯示建議
     */
    showSuggestions(suggestions, type = 'general') {
        if (!this.suggestionContainer || suggestions.length === 0) return;
        
        // 根據類型決定是否顯示
        if (type === 'realtime' && this.guideLevel === 'minimal') {
            return; // 最小幫助模式下不顯示實時提示
        }
        
        // 創建提示元素
        const container = document.createElement('div');
        container.className = `realtime-guide ${type}`;
        
        suggestions.forEach(suggestion => {
            const item = document.createElement('div');
            item.className = `guide-item guide-${suggestion.severity || 'info'}`;
            
            const message = document.createElement('div');
            message.className = 'guide-message';
            message.textContent = suggestion.message;
            item.appendChild(message);
            
            if (suggestion.suggestion) {
                const suggestionText = document.createElement('div');
                suggestionText.className = 'guide-suggestion';
                suggestionText.textContent = suggestion.suggestion;
                item.appendChild(suggestionText);
            }
            
            if (suggestion.example) {
                const example = document.createElement('code');
                example.className = 'guide-example';
                example.textContent = suggestion.example;
                item.appendChild(example);
            }
            
            container.appendChild(item);
        });
        
        // 更新容器
        this.suggestionContainer.innerHTML = '';
        this.suggestionContainer.appendChild(container);
        
        // 自動隱藏（除了警告）
        if (type !== 'guidance' || suggestions.every(s => s.severity !== 'warning')) {
            setTimeout(() => {
                if (container.parentElement) {
                    container.style.opacity = '0';
                    setTimeout(() => {
                        if (container.parentElement) {
                            container.remove();
                        }
                    }, 300);
                }
            }, 5000);
        }
    }
    
    /**
     * 設置幫助級別
     */
    setGuideLevel(level) {
        this.guideLevel = level;
        // 可以根據級別調整延遲時間
        if (level === 'minimal') {
            this.hintDelay = 5000; // 5秒
        } else if (level === 'moderate') {
            this.hintDelay = 2000; // 2秒
        } else {
            this.hintDelay = 1000; // 1秒
        }
    }
    
    /**
     * 清除所有提示
     */
    clear() {
        if (this.suggestionContainer) {
            this.suggestionContainer.innerHTML = '';
        }
    }
}

// 導出供全局使用
if (typeof module !== 'undefined' && module.exports) {
    module.exports = RealtimeGuide;
}

File: web_tutor/realtime_guide.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
實時引導系統
用於在用戶輸入過程中提供智能引導和提示
"""

import ast
import re
from typing import Dict, List, Tuple, Optional, Any


class RealtimeGuide:
    """實時引導系統，提供輸入過程中的智能提示"""
    
    def __init__(self, lesson: Optional[Dict[str, Any]] = None):
        """
        初始化實時引導系統
        
        Args:
            lesson: 當前課程信息，包含驗證器要求
        """
        self.lesson = lesson
        self.code_requirements = {}
        if lesson and 'validator' in lesson:
            self.code_requirements = lesson['validator'].get('code_requirements', {})
    
    def check_syntax(self, code: str) -> Tuple[bool, Optional[str], Optional[int]]:
        """
        檢查語法錯誤
        
        Returns:
            (是否有語法錯誤, 錯誤訊息, 錯誤行號)
        """
        try:
            ast.parse(code)
            return False, None, None
        except SyntaxError as e:
            return True, e.msg, e.lineno
        except Exception as e:
            return True, str(e), None
    
    def analyze_partial_code(self, code: str, cursor_line: int, cursor_col: int) -> Dict[str, Any]:
        """
        分析部分代碼，提供實時引導
        
        Args:
            code: 當前代碼
            cursor_line: 游標所在行（從1開始）
            cursor_col: 游標所在列（從1開始）
        
        Returns:
            包含引導信息的字典
        """
        suggestions = []
        warnings = []
        hints = []
        
        lines = code.split('\n')
        current_line = lines[cursor_line - 1] if cursor_line <= len(lines) else ""
        
        # 1. 檢查語法錯誤
        has_syntax_error, error_msg, error_line = self.check_syntax(code)
        if has_syntax_error and error_line:
            if error_line == cursor_line:
                suggestions.append({
                    "type": "syntax_error",
                    "message": f"語法錯誤：{error_msg}",
                    "line": error_line,
                    "severity": "error"
                })
        
        # 2. 根據課程要求提供引導
        if self.code_requirements:
            # 檢查是否需要循環
            if self.code_requirements.get("requires_loop", False):
                if not self._has_loop_in_code(code):
                    loop_type = self.code_requirements.get("loop_type", "")
                    if loop_type == "for":
                        if "for" not in code.lower():
                            hints.append({
                                "type": "missing_loop",
                                "message": "💡 提示：此題目要求使用 for 循環",
                                "suggestion": "嘗試使用：for i in range(...):",
                                "severity": "info"
                            })
                    elif loop_type == "while":
                        if "while" not in code.lower():
                            hints.append({
                                "type": "missing_loop",
                                "message": "💡 提示：此題目要求使用 while 循環",
                                "suggestion": "嘗試使用：while 條件:",
                                "severity": "info"
                            })
                    else:
                        if "for" not in code.lower() and "while" not in code.lower():
                            hints.append({
                                "type": "missing_loop",
                                "message": "💡 提示：此題目要求使用循環",
                                "suggestion": "可以使用 for 或 while 循環",
                                "severity": "info"
                            })
            
            # 檢查是否禁止硬編碼
            if self.code_requirements.get("forbids_hardcode", False):
                if self._detect_hardcode(code):
                    warnings.append({
                        "type": "hardcode_detected",
                        "message": "⚠️ 檢測到可能的硬編碼寫法",
                        "suggestion": "考慮使用循環來簡化代碼",
                        "severity": "warning"
                    })
            
            # 檢查是否需要函數
            if self.code_requirements.get("requires_function", False):
                if "def " not in code:
                    function_name = self.code_requirements.get("function_name", "")
                    if function_name:
                        hints.append({
                            "type": "missing_function",
                            "message": f"💡 提示：需要定義名為 '{function_name}' 的函數",
                            "suggestion": f"def {function_name}():",
                            "severity": "info"
                        })
                    else:
                        hints.append({
                            "type": "missing_function",
                            "message": "💡 提示：此題目要求定義函數",
                            "suggestion": "使用 def 函數名稱(): 來定義函數",
                            "severity": "info"
                        })
            
            # 檢查是否需要 if 語句
            if self.code_requirements.get("requires_if", False):
                if "if " not in code:
                    hints.append({
                        "type": "missing_if",
                        "message": "💡 提示：此題目要求使用 if 語句",
                        "suggestion": "使用 if 條件: 來進行條件判斷",
                        "severity": "info"
                    })
        
        # 3. 檢查常見錯誤模式
        common_errors = self._check_common_errors(code, current_line, cursor_line)
        suggestions.extend(common_errors)
        
        return {
            "suggestions": suggestions,
            "warnings": warnings,
            "hints": hints,
            "has_syntax_error": has_syntax_error
        }
    
    def _has_loop_in_code(self, code: str) -> bool:
        """檢查代碼中是否有循環"""
        try:
            tree = ast.parse(code)
            for node in ast.walk(tree):
                if isinstance(node, (ast.For, ast.While)):
                    return True
            return False
        except:
            return False
    
    def _detect_hardcode(self, code: str) -> bool:
        """檢測硬編碼模式"""
        lines = code.split('\n')
        print_count = sum(1 for line in lines if 'print(' in line)
        
        # 如果有多個 print 且沒有循環，可能是硬編碼
        if print_count > 3 and not self._has_loop_in_code(code):
            return True
        
        return False
    
    def _check_common_errors(self, code: str, current_line: str, cursor_line: int) -> List[Dict[str, Any]]:
        """檢查常見錯誤"""
        suggestions = []
        
        # 檢查縮排問題
        if current_line.strip() and not current_line.startswith(' ') and not current_line.startswith('\t'):
            # 檢查上一行是否以冒號結尾（需要縮排）
            lines = code.split('\n')
            if cursor_line > 1:
                prev_line = lines[cursor_line - 2].strip()
                if prev_line.endswith(':'):
                    suggestions.append({
                        "type": "indentation",
                        "message": "💡 提示：上一行以冒號結尾，這一行需要縮排",
                        "suggestion": "按 Tab 或輸入 4 個空格",
                        "severity": "info",
                        "line": cursor_line
                    })
        
        # 檢查常見的拼寫錯誤
        common_mistakes = {
            "prnt": "print",
            "prin": "print",
            "rng": "range",
            "rge": "range",
            "fr": "for",
            "whle": "while",
            "whil": "while",
            "if ": "if ",
            "el": "else",
            "els": "else",
            "elif": "elif"
        }
        
        for mistake, correct in common_mistakes.items():
            if mistake in current_line.lower() and correct not in current_line.lower():
                # 檢查是否可能是拼寫錯誤
                if re.search(rf'\b{mistake}\b', current_line, re.IGNORECASE):
                    suggestions.append({
                        "type": "typo",
                        "message": f"💡 是否想輸入 '{correct}'？",
                        "suggestion": f"將 '{mistake}' 改為 '{correct}'",
                        "severity": "hint",
                        "line": cursor_line
                    })
                    break
        
        return suggestions
    
    def get_contextual_suggestion(self, code: str, cursor_line: int, cursor_col: int) -> Optional[str]:
        """
        根據上下文提供建議
        
        Returns:
            建議文字或 None
        """
        lines = code.split('\n')
        if cursor_line > len(lines):
            return None
        
        current_line = lines[cursor_line - 1]
        
        # 如果正在輸入 for 循環
        if "for" in current_line and "in" not in current_line and "range" not in current_line:
            return "💡 提示：for 循環通常搭配 range() 使用，例如：for i in range(5):"
        
        # 如果正在輸入函數定義
        if "def" in current_line and ":" not in current_line:
            return "💡 提示：函數定義需要以冒號 : 結尾"
        
        # 如果正在輸入 if 語句
        if "if" in current_line and ":" not in current_line and ":" not in current_line[-3:]:
            return "💡 提示：if 語句需要以冒號 : 結尾"
        
        return None
    
    def should_show_hint(self, code: str, time_typing: float = 0) -> bool:
        """
        判斷是否應該顯示提示
        
        Args:
            code: 當前代碼
            time_typing: 輸入時間（秒）
        
        Returns:
            是否應該顯示提示
        """
        # 如果代碼很短，不顯示提示（讓學習者先思考）
        if len(code.strip()) < 10:
            return False
        
        # 如果輸入時間超過 5 秒還沒有進展，顯示提示
        if time_typing > 5:
            return True
        
        # 如果有語法錯誤，顯示提示
        has_error, _, _ = self.check_syntax(code)
        if has_error:
            return True
        
        return False


def create_guide_for_lesson(lesson: Dict[str, Any]) -> RealtimeGuide:
    """為課程創建引導系統"""
    return RealtimeGuide(lesson)

